### PR TITLE
Add windows default paths to lex tool

### PR DESCRIPTION
**Pull request by dmoody256**
I discovered this issue when testing #3270. Lex tool needs to use --nounistd flag on windows to prevent including 'unistd.h'.

This PR:
* adds default paths for the lex tool on windows
* adds chocolatey/bin as a default path to win32 platform
* adds check for 'win_flex' binary on windows
* adds --nounistd as default flag on windows in the LEXUNISTD construction variable
* updates lex tests to handle --nounistd
* add a test for making sure Environments can be constructed when no lex is found
* add live mingw test, to make sure 'unistd.h' can be used with lex tool in mingw environments
* add documentation for LEXUNISTD construction variable

## Contributor Checklist:

* [x] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [x] I have updated `master/src/CHANGES.txt` directory (and read the `README.txt` in that directory)
* [x] I have updated the appropriate documentation

### CODE REVIEW

File: .appveyor.yml [hunk 1]
```diff
@@ -17,7 +17,7 @@ install:
   # add python and python user-base to path for pip installs
   - cmd: "C:\\%WINPYTHON%\\python.exe --version"
   - cmd: for /F "tokens=*" %%g in ('C:\\%WINPYTHON%\\python.exe -m site --user-site') do (set PYSITEDIR=%%g)
-  - cmd: "set PATH=C:\\%WINPYTHON%;C:\\%WINPYTHON%\\Scripts;C:\\cygwin64\\bin;C:\\msys64;C:\\ProgramData\\chocolatey\\bin;%PATH%"
+  - cmd: "set PATH=C:\\%WINPYTHON%;C:\\%WINPYTHON%\\Scripts;C:\\MinGW\\bin;C:\\MinGW\\msys\\1.0\\bin;C:\\cygwin\\bin;C:\\ProgramData\\chocolatey\\bin;%PATH%"
```
Comment (bdbaddog) [inline .appveyor.yml]: Is msys64 an invalid path?

Comment (dmoody256) [reply .appveyor.yml]: yeah, the standard default path for msys64 is C:\msys64\usr\bin. We should only have 32 bit or 64 bit in the path, but not both.

Comment (bdbaddog) [reply .appveyor.yml]: so are we choosing 32 bit then with these changes?
Can you add a more explicit blurb to what path changes you're doing here and why in the summary? It's a bit difficult to tell via the diff and having to mentally reverse engineer the why doesn't make sense..

Comment (dmoody256) [reply .appveyor.yml]: It's a bit complicated why we need 32 bit so bare with me. Looking at the appveyor docs, we should be using 64 bit, because it is on both the 2015 and 2017 appveyor images. 

However, the problem with 64 bit is that the mingw tool will pick up the default 32 bit version first in its list of paths and build with that, and then the Win32/mingw.py test will fail when it tries to execute this:
https://github.com/SCons/scons/blob/9f090132091e56c123262106f3563142787c2fcf/test/Win32/mingw.py#L156
because the TestCmd run process will have the os.environ path which will have 64 bit bins in it:
https://github.com/SCons/scons/blob/9f090132091e56c123262106f3563142787c2fcf/testing/framework/TestCmd.py#L1428-L1433
Using 32 bit it will work, and skip on the 2017 image because the test framework wont find mingw, example recent appveyor 2017 build:
https://ci.appveyor.com/project/SCons/scons/builds/22156672/job/mnvvvpgodtsacjpj#L274

So to summarize we need 32 bit until we have the #3291 implemented and update Win32/mingw.py test to select a version based on whats in the the os.environ PATH.

File: test/LEX/live.py [hunk 2]
```diff
@@ -40,15 +42,24 @@
...
     test.skip_test('No lex or flex found; skipping test.\n')
 
+tools = "'default'"
+if sys.platform == 'win32':
+    # make sure mingw is installed on win32
+    if not test.where_is('gcc'):
+        test.skip_test('No mingw on windows; skipping test.\n')
+    # lex on win32 has a dependencies on mingw for unix headers
+    # so add it as a tool to the environment.
+    tools += ", 'mingw'"
```
Comment (bdbaddog) [inline test/LEX/live.py]: Should you still need this with the logic to check mingw paths in lex tool above?

Comment (dmoody256) [reply test/LEX/live.py]: The Unix headers won't be part of the default include path unless mingw is added to the environment on windows. I don't know exactly how lex works, but this test generates some code which needs Unix headers which mingw brings to windows.

Comment (bdbaddog) [reply test/LEX/live.py]: lex takes an input file which specifies tokens to be tokenized as part a parser (often build with yacc/bison/byacc).  It generates a .c and a .h file (and potentially other debug output files).

It should be able to be compiled by MSVC on windows though. 
Does it fail with msvc?

**Comment by dmoody256** [comment]
Yeah, it fails to compile because it can't find unistd.h with just the
default tools on windows.

File: test/LEX/live.py (hunk 2 above)
Comment (dmoody256) [reply test/LEX/live.py]: Yeah, it fails to compile because it can't find unistd.h with just the default tools on windows.

Do you know if it possible to use lex to generate code that doesn't include unix stuff? I will take a look when I get a chance.

Comment (bdbaddog) [reply test/LEX/live.py]: It should be able to. Which lex are you picking up in the tests?
Maybe try:
https://sourceforge.net/projects/winflexbison/

Which is in chocolatey.

Basically the skeleton file used by flex/lex will have unistd if flex is built in an environment which has it. so likely if you're using mingw's lex it will depend on unistd. Also you can manually specify the skeleton file. (Using -Sskeleton)

Comment (bdbaddog) [reply test/LEX/live.py]: Also in the past I've seen  sed scripts to rip the #include unistd.h out.

Comment (mwichmann) [reply test/LEX/live.py]: flex at least has an option --nounistd

Comment (mwichmann) [reply test/LEX/live.py]: And this for winflexbison:  https://github.com/lexxmark/winflexbison/commit/daca9cadd0f7139a9875f417ddf8872cfb1f338f

Comment (dmoody256) [reply test/LEX/live.py]: ok added a check for 'win_flex' binary and used the --nounistd on windows as default flag. That flag also works with flex in mingw.

File: src/engine/SCons/Tool/lex.py [hunk 3]
```diff
@@ -74,27 +75,26 @@ def get_lex_path(env, append_paths=False):
     """
     # save existing path to reset if we don't want to append any paths
     envPath = env['ENV']['PATH']
-    bins = ['lex', 'flex']
+    bins = ['win_flex', 'lex', 'flex']
```
Comment (bdbaddog) [inline src/engine/SCons/Tool/lex.py]: SHould win_flex only be searched for on win32?

Comment (dmoody256) [reply src/engine/SCons/Tool/lex.py]: Currently this function is only ever called in win32 case, and the default paths also represent that, they all use paths with 'C:\'. In non win32 case, the system defaults from other platforms add some paths:  https://github.com/SCons/scons/blob/44740370071393c3de04a456506b9f2617350e05/src/engine/SCons/Platform/posix.py#L94

and then env.Detect method locates flex or lex in this PR

File: src/CHANGES.txt [hunk 4]
```diff
@@ -19,6 +19,9 @@ RELEASE  VERSION/DATE TO BE FILLED IN LATER
     - Change the default for AppendENVPath to delete_existing=0, so path
       order will not be changed, unless explicitly set (Issue #3276)
     - Fixed bug which threw error when running SCons on windows system with no MSVC installed.
+    - Update link tool to convert target to node before accessing node member
+    - Update mingw tool to remove MSVC like nologo CCFLAG
+    - Update lex tool to work in mingw environments
```
Comment (bdbaddog) [inline src/CHANGES.txt]: Add that it's going to search for win_flex on windows? (And that it's given preference.

Comment (dmoody256) [reply src/CHANGES.txt]: added more info to CHANGES.txt in 362aae2

File: src/engine/SCons/Tool/lex.py (hunk 3 above)
Comment (bdbaddog) [reply src/engine/SCons/Tool/lex.py]: So the reason to search in that order is to not pick up the mingw or cygwin flex/lex?

Comment (dmoody256) [reply src/engine/SCons/Tool/lex.py]: yeah on windows, you would prefer to use win_flex which is from chocolaty, over flex/lex from cygwin/mingw. win_flex will work the same for mingw environments as flex/lex, which the new live_mingw.py tests. I also tested independently mingw flex on windows environment, but that environment setup is not tested in the new tests I added. I suppose I could mock out the CHOCO_DEFAULT_PATH so that win_flex is not found and it would skip over to flex in an additional mingw test for our unit tests.

Comment (bdbaddog) [reply src/engine/SCons/Tool/lex.py]: I think you're missing what I'm saying. So if a user on windows has been using flex from mingw and also has win_flex installed, when they pull this update they'll get a different flex. So minimally we need to highlight this in the CHANGES.txt.  Though optimally we'd just add it to the end of the list.  If you're looking to make the tests work on the test infrastructure should be able to set LEX='win_flex' in the tests when on windows?

The key item is. We don't change things without telling the users in CHANGES.txt, and ideally we don't change things just so our ci environment works which could break in the users environment which worked with previous SCons.  Nothing more infuriating to a user than pulling and update, having your build break, no info in the changelog, and you have to go search sources to figure out what happened.. Then you send flame to the devs (rightfully so)..

Comment (dmoody256) [reply src/engine/SCons/Tool/lex.py]: I didn't set the order to make any tests work. It works either way. I chose win_flex as the first default because if your on windows that seems like an appropriate default. I'll move it to the back to preserve existing defaults. 

In regards to the case you mentioned however, existing builds would be different with this PR even if win_flex is at the back of the list because they would be using the --no-unistd option in mingw builds now. So the existing default would require unistd.h on windows which is only possible in posix environment on windows like msys/cygwin, but would be broken in environments that don't support posix. This PR changes the default on windows to use --no-unistd option which works in either case, but will be different than before. Is that going to be an issue?

Comment (bdbaddog) [reply src/engine/SCons/Tool/lex.py]: do you need --no-unistd if using mingw's gcc? or only if using MSVC ?

Comment (dmoody256) [reply src/engine/SCons/Tool/lex.py]: you don't need it with mingw, as it supports unistd.h, but it works with it as well. MSVC doesn't have unistd.h by default, so you need it.

Comment (bdbaddog) [reply src/engine/SCons/Tool/lex.py]: So from looking at the logic if your on win32 it's always going to have --nounistd? (which should really only be in place if not using cygwin or mingw?)

Comment (dmoody256) [reply src/engine/SCons/Tool/lex.py]: I agree that would be a better default, but what would be a good way to know if the user intends to setup a posix environment on windows? I thought about checking for these during lex generation, but then it would depend on the order you loaded tools?

Comment (bdbaddog) [reply src/engine/SCons/Tool/lex.py]: I'm guessing we'd be right the majority of the time if we checked for MS VC and only set -nounistd when that's being used.  We don't have clang-cl tool yet.   Maybe LEXNOUNISTD could be set in MSVC tool and if that's set then add it? That way tools can turn it on per tool, and users can request it as well?

Comment (dmoody256) [reply src/engine/SCons/Tool/lex.py]: ok updated with MSVC tool setting no-unistd in 7b192e7

File: src/engine/SCons/Tool/msvc.py [hunk 5]
```diff
@@ -271,6 +271,10 @@ def generate(env):
     env['SHOBJPREFIX']    = '$OBJPREFIX'
     env['SHOBJSUFFIX']    = '$OBJSUFFIX'
 
+    # MSVC probably wont support unistd.h so default
+    # without it for lex generation
+    env["LEXUNISTD"] = SCons.Util.CLVar("--nounistd")
```
Comment (bdbaddog) [inline src/engine/SCons/Tool/msvc.py:276]: Looks like there's a flag "--wincompat" which should be used instead?

Comment (dmoody256) [reply src/engine/SCons/Tool/msvc.py:276]: --wincompat is only available in win_flex I believe, while --nounistd is in both. 

--wincompat is a bit better for MSVC, as it will replace some of things unistd.h brings in.

From win_flex [Changelog](https://github.com/lexxmark/winflexbison/blob/master/changelog.md):
```
* added --wincompat option to win_flex (this option changes `<unistd.h>` unix include with `<io.h>` windows analog
  also `isatty/fileno` functions changed to `_isatty/_fileno`)
```

Any ideas how to detect if win_flex will be used from msvc.py? I assume the user would need to set this if the generated source needs something from the <io.h> include. Our tests don't cover those cases either.

**Comment by bdbaddog** [comment]
I'm going to merge and maybe update the flag for windows per my comment above if that works as well. Thanks for your diligence in this PR.

**Comment by bdbaddog** [comment]
Ugh.. Yes this kinda thing is part of the driver for a tools/platform
rearch that we've talked about a few times but hasn't moved forward very
far.
Worst case check LEX string in the lex tool in the if win32 if clause. :(

-Bill
